### PR TITLE
When all submissions are deleted for a source, delete the enclosing directory

### LINKED ISSUE #1474: Local safe-delete removes enclosing directory, remote delete does not
## Description

Reported by @cfm:  

Local 'safe-delete' (deleting files and messages) also deletes the directory containing the file and message data (`~/.securedrop_client/data/<journalist-designator>`), while remote delete of files and messages removes the content but leaves the directory in place.
 
## Steps to Reproduce
- Log into the JI and delete source A's files and messages
- Locally delete source B's files and messages 
- Inspect `~/.securedrop_client/data/` directory and observe source A directory (now empty) is still present

## Expected Behavior
- Client and server deletion behave the same

## Actual Behavior
- Source A directory is present, source B directory is gone

## Comments

I kind of prefer the client deletion behaviour, because it really cleans up _all_ of the indications that messages and files were downloaded, vs removing the files but leaving the directory in place.

### PR DESCRIPTION
**Pull request by rocodes**
# Description

Fixes #1474

# Test Plan
- [ ] Run through STR in the linked issue and confirm safe deletion behaviour on sync (deletion of files and messages via JI) and deletion behaviour via local deletion produce the same results--the enclosing directory is removed when empty
- Choose a source with multiple file submissions. Download only one file via the client, and delete only that one submission via the JI, leaving the others intact but not yet downloaded.
- [ ] The enclosing directory is removed
- [ ] The enclosing directory is recreated when another submission is downloaded

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [x] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

### CODE REVIEW

File: securedrop_client/storage.py [hunk 1]
```diff
@@ -940,6 +954,24 @@ def delete_single_submission_or_reply_on_disk(
...
 
+def _cleanup_directory_if_empty(target_dir: str) -> None:
+    """
+    If a directory is empty, remove it. Used after sync to ensure that an empty
+    directory is not left behind after files and messages are deleted.
+    """
+
+    try:
+        if os.path.isdir(target_dir):
+            if not next(os.scandir(target_dir), None):
```
Comment (cfm) [inline securedrop_client/storage.py:965]: Very efficient to check `scandir()` rather than `len() == 0` on `os.listdir()` or `os.walk()`; nice!

**Review by cfm — APPROVED** [review]
Test plan checks out!  Approving, with nits, based on my requested review; but deferring to @gonzalo-bulnes as the current assignee for the final review and merge.

Good catch, @rocodes, that this is really the case where _the source no longer has any downloaded files_, which is a broader case than that where _the source's conversation has just been safe-deleted_.

- [x] Run through STR in the linked issue and confirm safe deletion behaviour on sync (deletion of files and messages via JI) and deletion behaviour via local deletion produce the same results--the enclosing directory is removed when empty
- Choose a source with multiple file submissions. Download only one file via the client, and delete only that one submission via the JI, leaving the others intact but not yet downloaded.
- [x] The enclosing directory is removed
- [x] The enclosing directory is recreated when another submission is downloaded

File: tests/test_storage.py [hunk 2]
```diff
@@ -2026,6 +2031,91 @@ def test___update_submissions_skip_uuids_successful(mocker, session, homedir):
...
+    mock_error = mocker.MagicMock()
+    mocker.patch("securedrop_client.storage.logger.error", mock_error)
+
+    _cleanup_directory_if_empty(empty_dir)
+    _cleanup_directory_if_empty(nonempty_dir)
+    _cleanup_directory_if_empty(not_a_dir)
+
+    assert not os.path.exists(empty_dir)
+    assert os.path.exists(nonempty_dir)
+    mock_error.assert_called_once_with("Error: method called on missing or malformed directory")
```
[the remark this answers is not in the document]
Comment (gonzalo-bulnes) [reply tests/test_storage.py:2062]: In this case, I think I'd also lean towards interleaving the actions and assertions like @cfm suggests.

**Review by gonzalo-bulnes — APPROVED** [review]
Thanks @cfm or the review!

Since the open comments don't affect the code behavior, instead of waiting for @rocodes to be back, I'll add a commit to apply @cfm's suggestion (asusming I understood it correctly) and fix the typo and merge. If I got anything wrong, it will be easy enough to correct.

**Pull request merged**